Sources/Datadog/URLSessionAutoInstrumentation/DDURLSessionDelegate.swift

```diff
@@ -11,11 +11,21 @@ open class DDURLSessionDelegate: NSObject, URLSessionTaskDelegate {
     private let interceptor: URLSessionInterceptorType?
 
     @objc
-    override public convenience init() {
-        self.init(interceptor: URLSessionAutoInstrumentation.instance?.interceptor)
+    override public init() {
+        self.interceptor = URLSessionAutoInstrumentation.instance?.interceptor
+        if interceptor == nil {
+            let error = ProgrammerError(
+                description: """
+                To use `DDURLSessionDelegate` you must specify first party hosts in `Datadog.Configuration` -
+                use `track(firstPartyHosts:)` to define which requests should be tracked.
+                """
+            )
+            consolePrint("\(error)")
+        }
     }
 
     internal init(interceptor: URLSessionInterceptorType?) {
```

Comment:
We don't need this initializer now as it is only used in tests. We should remove it and mock the global dependency (`URLSessionAutoInstrumentation.instance?.interceptor`) in the same way as we do it for other features:
```swift
func test...() {
   URLSessionAutoInstrumentation.instance = .mockWith(interceptor: interceptor)
   defer { URLSessionAutoInstrumentation.instance = nil }

   let delegate = DDURLSessionDelegate() // it will use the `interceptor`

   // ...
}
```

Reply:
but we don't need a `URLSessionAutoInstrumentation` instance to create a delegate neither

i think
1. we can make `DDURLSessionDelegate.interceptor` an `internal var` so that we can change internally in unit tests
2. or we can keep duplicate init's (ideally we would have those methods anyway but only due to Swift initialization rules their implementations are not ideal)

wdyt?

Reply:
> but we don't need a `URLSessionAutoInstrumentation` instance to create a delegate neither

Sorry, I don't understand 🤔. Yes, we don't need it, but how does it relate to the test I proposed?

> we can make `DDURLSessionDelegate.interceptor` an `internal var` so that we can change internally in unit tests

Yes, this could be a way to go as well. If it's `internal`, we can also test that `interceptor` is set correctly when instrumentation is enabled:
```swift
func testGivenURLSessionAutoInstrumentationEnabled_whenCreatingDelegate_itUsesInterceptor() {
   // Given
   // ...

   // When
   let delegate = DDURLSessionDelegate()
   
   // Then
   XCTAssertTrue(delegate.interceptor === URLSessionAutoInstrumentation.instance.interceptor)
}
```
Right now, this explicit test was not possible.

> or we can keep duplicate init's (ideally we would have those methods anyway but only due to Swift initialization rules their implementations are not ideal)

We should avoid doing this if necessary. We only do it when other ways are not possible (AFAIK there are only 2, carefully controlled places like this).